src/Access/LDAPAccessStorage.h

```diff
@@ -0,0 +1,68 @@
+#pragma once
+
+#include <Access/MemoryAccessStorage.h>
+#include <Core/Types.h>
+#include <ext/scope_guard.h>
+#include <mutex>
+#include <set>
+
+
+namespace Poco
+{
+    namespace Util
+    {
+        class AbstractConfiguration;
+    }
+}
+
+
+namespace DB
+{
+class AccessControlManager;
+
+/// Implementation of IAccessStorage which allows attaching users from a remote LDAP server.
+/// Currently, any user name will be treated as a name of an existing remote user,
+/// a user info entity will be created, with LDAP_SERVER authentication type.
+class LDAPAccessStorage : public IAccessStorage
+{
+public:
+    static constexpr char STORAGE_TYPE[] = "ldap";
+
+    explicit LDAPAccessStorage(const String & storage_name_, AccessControlManager * access_control_manager_, const Poco::Util::AbstractConfiguration & config, const String & prefix);
+    virtual ~LDAPAccessStorage() override = default;
+
+public: // IAccessStorage implementations.
+    virtual const char * getStorageType() const override;
+    virtual String getStorageParamsJSON() const override;
+
+private: // IAccessStorage implementations.
+    virtual std::optional<UUID> findImpl(EntityType type, const String & name) const override;
+    virtual std::vector<UUID> findAllImpl(EntityType type) const override;
+    virtual bool existsImpl(const UUID & id) const override;
+    virtual AccessEntityPtr readImpl(const UUID & id) const override;
+    virtual String readNameImpl(const UUID & id) const override;
+    virtual bool canInsertImpl(const AccessEntityPtr &) const override;
+    virtual UUID insertImpl(const AccessEntityPtr & entity, bool replace_if_exists) override;
+    virtual void removeImpl(const UUID & id) override;
+    virtual void updateImpl(const UUID & id, const UpdateFunc & update_func) override;
+    virtual ext::scope_guard subscribeForChangesImpl(const UUID & id, const OnChangedHandler & handler) const override;
+    virtual ext::scope_guard subscribeForChangesImpl(EntityType type, const OnChangedHandler & handler) const override;
+    virtual bool hasSubscriptionImpl(const UUID & id) const override;
+    virtual bool hasSubscriptionImpl(EntityType type) const override;
+    virtual UUID loginImpl(const String & user_name, const String & password, const Poco::Net::IPAddress & address, const ExternalAuthenticators & external_authenticators) const override;
```

Comment:
I think it's necessary to implement `getIDOfLoggedUserImpl()` in `LDAPAccessStorage`, otherwise `LDAPAccessStorage` wouldn't work when the interserver secret is enabled.

Reply:
The default implementation of `getIDOfLoggedUserImpl()` calls `getID()`, which in its turn calls `findImpl()`, which is implemented in `LDAPAccessStorage` and just forwards call to internal `MemoryAccessStorage` "cache".

Reply:
Yes, I know. And that's the reason why it won't work.
When the interserver secret is enabled an user is authorized on one node, and then while processing a distributed query other nodes are kind of pretending that the user is authorized there too, without passing actual credentials. The problem is that the internal `MemoryAccessStorage` inside `LDAPAccessStorage` won't contain an user which has authorized on other node.

Reply:
So the custom implementation must either find an existing user in the cache, or create one as it does in `loginImpl()` but assume it is authenticated? OK.

Reply:
Pushed.

Reply:
Maybe I should write a better comment for `getIDOfLoggerUser()`, but it's quite difficult to explain what this function is for.

Reply:
The confusion here comes more from the fact that `LDAPAccessStorage` creates users in `loginImpl()`, so `getIDOfLoggerUser()` should also do that.